src/transformers/trainer.py

```diff
@@ -2202,9 +2238,15 @@ def _save(self, output_dir: Optional[str] = None, state_dict=None):
                 logger.info("Trainer.model is not a `PreTrainedModel`, only saving its state dict.")
                 if state_dict is None:
                     state_dict = self.model.state_dict()
-                torch.save(state_dict, os.path.join(output_dir, WEIGHTS_NAME))
+                if is_sagemaker_mp_enabled():
+                    smp.save(state_dict, os.path.join(output_dir, WEIGHTS_NAME), partial=True)
+                else:
+                    torch.save(state_dict, os.path.join(output_dir, WEIGHTS_NAME))
         else:
-            self.model.save_pretrained(output_dir, state_dict=state_dict)
+            if is_sagemaker_mp_enabled():
+                smp.save(state_dict, os.path.join(output_dir, WEIGHTS_NAME), partial=True)
+            else:
+                self.model.save_pretrained(output_dir, state_dict=state_dict)
```

Comment:
No calling `save_pretrained` here means the config will not be saved and the checkpoint won't be able to be loaded with `from_pretrained` independently of the training. It's not a regular checkpoint anyway, so maybe it's okay. Flagging this here anyway.

Reply:
SMP checkpoints are saved partially hence we do not want to shard SMP checkpoints. In order to use `save_pretrained` for SMP, we need to skip  `shard_checkpoint` for SMP. Independent of`max_shard_size` `shard_checkpoint` is called and we hit errors in `shard_checkpoint` since SMP checkpoints are different. If `shard_checkpoint` can be optional in `save_pretrained`, we can use `save_pretrained`  with `save_function=smp.save`. [In my previous PR](https://github.com/huggingface/transformers/pull/16734/files#diff-6b72b98c4c2dcfc6cc606843917733f5d858374fbc22a735ff483bbc0c1e63ea) I tried to skip `shard_checkpoint` for SMP, but feedback was not change `save_pretrained`.
 
`from_pretrained` won't work for SMP models. We are working on how to support fine-tuning. In this PR, I added support partial checkpoint saving/loading during training.